core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -449,21 +485,42 @@ private static String entryDefaultValue(Parameter entry) {
     @Nullable
     private static Object convertParameter(@Nullable String value, Parameter entry) {
         return AnnotatedHttpServiceParamUtil.convertParameter(value,
-                                                              entry.type,
-                                                              entry.enumConverter,
+                                                              entry.type(),
+                                                              entry.enumConverter(),
                                                               entry.isOptionalWrapped());
     }
 
     @Nullable
-    private static String httpHeaderValue(Parameter entry, HttpRequest req) {
+    private static Object httpHeaderValue(Parameter entry, HttpRequest req) {
         final String name = entry.name();
         assert name != null;
-        final String value = req.headers().get(AsciiString.of(name));
-        if (value != null) {
-            return value;
+
+        final Class<?> wrapperType = entry.wrapperType();
+        if (wrapperType != null && List.class.isAssignableFrom(wrapperType)) {
+            assert entry.type() == String.class;
```

Comment:
Could we append ` : entry.type().getName()`?

Reply:
Do you mean `assert entry.type().getName().equals(String.class.getName())`?